File: ios/Classes/lib/awesome_notifications_core/managers/DismissedManager.swift
//
//  DismissedManager.swift
//  awesome_notifications
//
//  Dismissed by CardaDev on 31/01/22.
//

import Foundation

public class DismissedManager {
    
    static let shared:SharedManager = SharedManager(tag: Definitions.SHARED_DISMISSED)
    
    public static func removeDismissed(id:Int) -> Bool {
        return shared.remove(referenceKey: String(id));
    }

    public static func listDismissed() -> [ActionReceived] {
        var returnedList:[ActionReceived] = []
        let dataList = shared.getAllObjects()
        
        for data in dataList {
            let received:ActionReceived = ActionReceived(nil).fromMap(arguments: data) as! ActionReceived
            returnedList.append(received)
        }
        
        return returnedList
    }

    public static func saveDismissed(received:NotificationReceived) {
        shared.set(received.toMap(), referenceKey: String(received.id!))
    }

    public static func getDismissedByKey(id:Int) -> ActionReceived? {
        guard let data:[String:Any?] = shared.get(referenceKey: String(id)) else {
          return nil
        }
        return ActionReceived(nil).fromMap(arguments: data) as? ActionReceived
    }

    public static func removeAllDismissed() {
        shared.removeAll()
    }

    public static func removeDismissed(id:Int) {
        _ = shared.remove(referenceKey: String(id))
    }
    
}

File: example/android/awn_core/src/main/java/me/carda/awesome_notifications/core/managers/DismissedManager.java
package me.carda.awesome_notifications.core.managers;

import android.content.Context;

import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.Calendar;
import java.util.List;

import me.carda.awesome_notifications.core.Definitions;
import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.models.returnedData.ActionReceived;
import me.carda.awesome_notifications.core.models.returnedData.NotificationReceived;
import me.carda.awesome_notifications.core.utils.StringUtils;

public class DismissedManager extends EventManager {

    private static final RepositoryManager<ActionReceived> shared
            = new RepositoryManager<>(
                    StringUtils.getInstance(),
                    "DismissedManager",
                    ActionReceived.class,
                    "ActionReceived");

    // ************** SINGLETON PATTERN ***********************

    private static DismissedManager instance;

    private DismissedManager(){}
    public static DismissedManager getInstance() {
        if (instance == null)
            instance = new DismissedManager();
        return instance;
    }

    // ************** SINGLETON PATTERN ***********************



    public boolean saveDismissed(
            @NonNull Context context,
            @NonNull ActionReceived received
    ) throws AwesomeNotificationsException {
        return shared.set(
                context,
                Definitions.SHARED_DISMISSED,
                _getKeyByIdAndDate(received.id, received.dismissedDate),
                received);
    }

    public List<ActionReceived> listDismisses(
            @NonNull Context context
    ) throws AwesomeNotificationsException {
        return shared.getAllObjects(context, Definitions.SHARED_DISMISSED);
    }

    public List<ActionReceived> getDismissedByKey(
            @NonNull Context context,
            @NonNull Integer id
    ) throws AwesomeNotificationsException {
        return shared.getAllObjectsStartingWith(
                context,
                Definitions.SHARED_DISMISSED,
                _getKeyById(id));
    }

    public NotificationReceived getDismissedByIdAndDate(
            @NonNull Context context,
            @NonNull Integer id,
            @NonNull Calendar dismissedDate
    ) throws AwesomeNotificationsException {
        return shared.get(
                context,
                Definitions.SHARED_DISMISSED,
                _getKeyByIdAndDate(id, dismissedDate));
    }

    public boolean removeDismissed(
            @NonNull Context context,
            @NonNull Integer id,
            @NonNull Calendar dismissedDate
    ) throws AwesomeNotificationsException {
        return shared.remove(
                context,
                Definitions.SHARED_DISMISSED,
                _getKeyByIdAndDate(id, dismissedDate));
    }

    public boolean removeAllDismissed(
            @NonNull Context context
    ) throws AwesomeNotificationsException {
        return shared.removeAll(
                context,
                Definitions.SHARED_DISMISSED);
    }

    public boolean commitChanges(
            @NonNull Context context
    ) throws AwesomeNotificationsException {
        return shared.commit(context);
    }
}
